### PR TITLE
glTF blocks

### PR DESCRIPTION
**Pull request by guusw**
Adding glTF loader blocks (#188 )

- [x] Static from file
- [x] Dynamic from binary
- [x] Dynamic from file

### CODE REVIEW

File: src/tests/gfx_gltf.edn [hunk 1]
```diff
@@ -0,0 +1,34 @@
+(def timestep (/ 1.0 120.0))
+(def Root (Node))
```
Comment (Kryptos-FR) [inline src/tests/gfx_gltf.edn:2]: You can use `(defnode root)`. In most samples (in documentation), we use lowercase names but we haven't really defined a convention yet.

File: src/tests/gfx_gltf.edn [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+(defn get-view-transform [time]
+  (->
+   (Float4 0 0.2 0.2 1.0) >= .arm-pos
+   time (Math.AxisAngleY) (Math.Rotation) (Math.MatMul .arm-pos) (ToFloat3) >= .arm-pos-3
+   {:Position .arm-pos-3 :Target (Float3 0 0.025 0)} (Math.LookAt)))
```
Comment (Kryptos-FR) [inline src/tests/gfx_gltf.edn:8]: Could use `(defblocks get-view-transform [time])` to avoid the indentation with `(->)`.

Is it completely equivalent though. @sinkingsugar I seem to remember that you told me that since it uses "chainification" there might be some corner cases or performance difference?

File: src/tests/gfx_gltf.edn [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+   (GFX.BuiltinFeature :Id BuiltinFeatureId.Transform) >> .features
+   (GFX.BuiltinFeature :Id BuiltinFeatureId.BaseColor) >> .features
+   (GFX.DrawablePass :Features .features) >> .render-steps
+
+   (GFX.glTF "external/glTF-Sample-Models/2.0/Avocado/glTF-Binary/Avocado.glb") >> .drawables
+
+    ;; Create view
+   {:Position (Float3 0 15 10) :Target (Float3 0 0 0)} (Math.LookAt) >= .view-transform
+   (GFX.View :View .view-transform) >= .view)
+  (GFX.MainWindow
```
Comment (Kryptos-FR) [inline src/tests/gfx_gltf.edn:22]: empty line after `(Setup)` makes it more readable.

File: src/tests/gfx_gltf.edn (hunk 2 above)
Comment (sinkingsugar) [reply src/tests/gfx_gltf.edn:8]: well `defblocks` is a macro so it's the same thing exactly I think.

File: src/extra/gfx/gltf.cpp [hunk 4]
```diff
@@ -13,12 +15,14 @@ namespace gfx {
...
-  static CBTypesInfo inputTypes() { return CoreInfo::AnyType; }
+  static inline Type TransformVarType = Type::VariableOf(CoreInfo::Float4x4Type);
+  static CBTypesInfo inputTypes() { return CoreInfo::AnyTableType; }
   static CBTypesInfo outputTypes() { return Types::DrawableHierarchy; }
 
   static CBParametersInfo parameters() {
     static Parameters params{
-        {"StaticPath", CBCCSTR("The static path to load a model from during warmup"), {CoreInfo::StringType}},
+        {"Transform", CBCCSTR("The transform variable to use"), {TransformVarType}},
+        {"Path", CBCCSTR("The static path to load a model from during warmup"), {CoreInfo::StringType}},
```
Comment (sinkingsugar) [inline src/extra/gfx/gltf.cpp:25]: Nice idea. As warmup can run from a worker thread before injecting it into the render loop thread.

File: src/extra/gfx/gltf.cpp [hunk 5]
```diff
@@ -13,12 +15,14 @@ namespace gfx {
...
   static inline Type ByteInputType = CoreInfo::BytesType;
-  static CBTypesInfo inputTypes() { return CoreInfo::AnyType; }
+  static inline Type TransformVarType = Type::VariableOf(CoreInfo::Float4x4Type);
+  static CBTypesInfo inputTypes() { return CoreInfo::AnyTableType; }
   static CBTypesInfo outputTypes() { return Types::DrawableHierarchy; }
 
   static CBParametersInfo parameters() {
     static Parameters params{
-        {"StaticPath", CBCCSTR("The static path to load a model from during warmup"), {CoreInfo::StringType}},
+        {"Transform", CBCCSTR("The transform variable to use"), {TransformVarType}},
```
Comment (sinkingsugar) [inline src/extra/gfx/gltf.cpp:24]: Is this to override/fix the default transform basically?

File: src/tests/gfx_gltf.edn (hunk 3 above)
Comment (guusw) [reply src/tests/gfx_gltf.edn:26]: Moved to a def and added newlines after Setups

File: src/extra/gfx/gltf.cpp (hunk 5 above)
Comment (guusw) [reply src/extra/gfx/gltf.cpp:24]: The same way as we decided it would work with the normal Drawable.
Where if you pass (GLTF :Transform <variable>) it will capture the variable and move the drawable every time you change the variable.

File: src/gfx/mesh.cpp [hunk 6]
```diff
@@ -1,46 +1,66 @@
...
+#include "math.hpp"
 #include <cassert>
 #include <magic_enum.hpp>
 #include <spdlog/spdlog.h>
 
 namespace gfx {
 
+size_t MeshFormat::computeVertexSize() const {
+  size_t vertexSize = 0;
+  for (auto &attr : vertexAttributes) {
```
Comment (inabe-a) [inline src/gfx/mesh.cpp:12]: [nit] const + few other places below

File: src/gfx/mesh.hpp [hunk 7]
```diff
@@ -37,6 +37,8 @@ struct MeshFormat {
   IndexFormat indexFormat = IndexFormat::UInt16;
   std::vector<MeshVertexAttribute> vertexAttributes;
 
+  size_t computeVertexSize() const;
```
Comment (inabe-a) [inline src/gfx/mesh.hpp:40]: getVertexSize() is more common

File: src/gfx/CMakeLists.txt [hunk 8]
```diff
@@ -40,6 +40,9 @@ target_precompile_headers(gfx PUBLIC "$<$<COMPILE_LANGUAGE:CXX>:${CMAKE_CURRENT_
 add_library(gfx_texture_file texture_file/texture_file.cpp)
 target_link_libraries(gfx_texture_file gfx)
 
+add_library(gfx_gltf gltf/gltf.cpp)
+target_link_libraries(gfx_gltf gfx tinygltf)
```
Comment (inabe-a) [inline src/gfx/CMakeLists.txt:44]: Good practice :) I'm also learning modern CMake recently. You probably already know this, but I learned a few tips from this guy: https://www.youtube.com/watch?v=bsXLMQ6WgIk

File: src/gfx/shader/wgsl_mapping.hpp [hunk 9]
```diff
@@ -59,16 +59,16 @@ inline String getFieldWGSLTypeName(const FieldType &type) {
   const char *baseType = nullptr;
   switch (type.baseType) {
   case ShaderFieldBaseType::UInt8:
-    baseType = "u8";
+    baseType = "u32";
```
Comment (inabe-a) [inline src/gfx/shader/wgsl_mapping.hpp:62]: Why is the base 32??

File: src/gfx/mesh.hpp (hunk 7 above)
Comment (guusw) [reply src/gfx/mesh.hpp:40]: Named it compute since it does some extra work than just returning a cached size

File: src/gfx/math.hpp [hunk 10]
```diff
@@ -11,6 +11,18 @@ inline float degToRad(float v) { return v * degToRadFactor; }
 
 constexpr float radToDegFactor = (180.0f / pi);
 inline float radToDeg(float v) { return v * radToDegFactor; }
+
+inline size_t alignTo(size_t size, size_t alignment) {
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/gfx/math.hpp:15]: I guess you mean `Round up to the next highest power of 2` from that article. Although I believe that one doesn't allow you to specify the alignment.
Good to know in case we need something similar though

File: src/gfx/shader/wgsl_mapping.hpp (hunk 9 above)
Comment (guusw) [reply src/gfx/shader/wgsl_mapping.hpp:62]: WebGPUdoesn't support anything useful on anything less than <u32/i32 in wgsl 
So when defining a field it's defined as u32 in the shader, but it might come from an uint8 vertex attribute in memory

References:
https://www.w3.org/TR/WGSL/#integer-types
https://www.w3.org/TR/webgpu/#enumdef-gpuvertexformat

File: src/gfx/drawable.hpp [hunk 11]
```diff
@@ -10,6 +10,8 @@
 
 namespace gfx {
 struct Drawable {
+  typedef std::shared_ptr<Drawable> PtrType;
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/gfx/drawable.hpp:13]: What do you mean? Using raw pointers? Naming it Ptr instead of PtrType?

File: src/gfx/math.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (inabe-a) [reply src/gfx/math.hpp:15]: Maybe it wasn't there after all.
size = (size + alignment - 1) & ~(alignment - 1) should work for powers of 2. (double check the math if you use it though)

File: src/gfx/shader/wgsl_mapping.hpp (hunk 9 above)
Comment (inabe-a) [reply src/gfx/shader/wgsl_mapping.hpp:62]: I see, I thought it's for the attributes. Ok, makes sense. They'll add support for fp16 at some point, but I think it's disabled by default unless you compile with that option.

File: src/gfx/drawable.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (inabe-a) [reply src/gfx/drawable.hpp:13]: I was suggesting name it Ptr, like in shared_ptr, it's easier to type etc.. Don't use raw pointers :)

File: src/gfx/tests/data.hpp [hunk 12]
```diff
@@ -1,7 +1,7 @@
-#ifndef GFX_TESTS_TEST_DATA
-#define GFX_TESTS_TEST_DATA
+#ifndef GFX_TESTS_DATA
```
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/gfx/tests/data.hpp:1]: lol https://stackoverflow.com/a/32693983

[the remark this answers is not in the document]
Comment (guusw) [reply src/gfx/tests/data.hpp:1]: https://marketplace.visualstudio.com/items?itemName=akiramiyakoda.cppincludeguard
This extension works pretty well, I'm thinking of using the GUID's instead of the filenames too since it's even easier with renames

[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/gfx/tests/data.hpp:1]: guid is a bit silly/naïve though imho :)

[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/gfx/tests/data.hpp:1]: but yeah collisions might be an issue at some point, although you likely notice immediately so the problem is not worth a solution ;)

**Review by Kryptos-FR — APPROVED** [review]

File: src/gfx/drawable.cpp [hunk 13]
```diff
@@ -0,0 +1,73 @@
+#include "drawable.hpp"
+#include <cassert>
+#include <set>
+
+namespace gfx {
+
+#define TRACK_VISITED 1
```
Comment (sinkingsugar) [inline src/gfx/drawable.cpp:7]: is this ok?

**Review by sinkingsugar — APPROVED** [review]

File: src/gfx/drawable.cpp (hunk 13 above)
Comment (guusw) [reply src/gfx/drawable.cpp]: Ah should've been and NDEBUG here somewhere. fixed that.
Also namespaced it

File: src/extra/gfx/drawable.cpp [hunk 14]
```diff
@@ -191,34 +185,35 @@ struct DrawShard : public BaseConsumer {
...
   }
 
-  SHVar activateSingle(SHContext *shContext, const SHVar &input) {
-    SHDrawable *shDrawable = (SHDrawable *)input.payload.objectValue;
-    updateSHDrawable(shDrawable);
-    addDrawableToQueue(shDrawable->drawable);
+  SHVar activateSeq(SHContext *shContext, const SHVar &input) {
+    auto &seq = input.payload.seqValue;
+    for (size_t i = 0; i < seq.len; i++) {
+      (void)activateSingle(shContext, seq.elements[i]);
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/extra/gfx/drawable.cpp:215]: The renderer internally groups draws using the same pipeline & mesh together and draws them as instances so we don't need a special path for it anymore

**Comment by sinkingsugar** [comment]
Codecov could be a red flag to check
